storage/client/src/main/java/com/azure/storage/file/FileAsyncClient.java

```diff
@@ -3,77 +3,494 @@
+        return azureFileStorageClient.files().startCopyWithRestResponseAsync(shareName, filePath, sourceUrl, null, metadata, Context.NONE)
+                    .map(this::startCopyResponse);
     }
 
+    /**
+     * Abort the copy.
+     * @param copyId
+     * @return
+     */
     public Mono<VoidResponse> abortCopy(String copyId) {
-        throw new UnsupportedOperationException();
+        return azureFileStorageClient.files().abortCopyWithRestResponseAsync(shareName, filePath, copyId, Context.NONE)
+                    .map(VoidResponse::new);
+    }
+
+    public Mono<Void> downloadToFile(String downloadFilePath) {
+        return downloadToFile(downloadFilePath, null);
+    }
+
+    public Mono<Void> downloadToFile(String downloadFilePath, FileRange range) {
+        AsynchronousFileChannel channel;
+        try {
+            channel = AsynchronousFileChannel.open(Paths.get(downloadFilePath), StandardOpenOption.READ, StandardOpenOption.WRITE);
+        } catch (IOException e) {
+            return Mono.error(e);
+        }
+        return sliceFileRange(range)
+                   .flatMap(chunk -> FluxUtil.bytebufStreamToFile(downloadWithProperties(chunk, false)
+                                                                      .flatMapMany(dar -> dar.value().body())
+                       , channel, chunk.start() - (range == null ? 0 : range.start()))
+                                         .doOnTerminate(() ->
+                                                            System.out.println("Saved " + chunk.toString() + " on thread " + Thread.currentThread().getName())))
+                   .then()
+                   .doOnTerminate(() -> {
+                       try {
+                           channel.close();
+                       } catch (IOException e) {
+                           throw new UncheckedIOException(e);
+                       }
+                   });
     }
 
-    public Mono<Response<FileDownloadInfo>> downloadWithProperties(long offset, long length, boolean rangeGetContentMD5) {
-        throw new UnsupportedOperationException();
+    private Flux<FileRange> sliceFileRange(FileRange fileRange) {
+        long offset = fileRange == null ? 0L : fileRange.start();
+        Mono<Long> end;
+        if (fileRange != null) {
+            end = Mono.just(fileRange.end());
+        } else {
+            end = Mono.empty();
+        }
+        end = end.switchIfEmpty(getProperties().map(rb -> rb.value().contentLength()));
+        return end
+                   .map(e -> {
+                       List<FileRange> chunks = new ArrayList<>();
+                       for (long pos = offset; pos < e; pos += FILE_DEFAULT_BLOCK_SIZE) {
+                           long count = FILE_DEFAULT_BLOCK_SIZE;
+                           if (pos + count > e) {
+                               count = e - pos;
+                           }
+                           chunks.add(new FileRange(pos, pos + count));
```

Comment:
On second thought, should the end be `pos + count - 1`?

Reply:
I agreed.
offset = 0; length = 7 
=>
pos = 0, end = 7
=>
count = 7
new FileRange(0, 7), but it is supposed to be new FileRange(0, 6).